clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -58,63 +58,43 @@
+    private int updateVersion;  // bumped on every metadata response
+    private int requestVersion; // bumped on every new topic addition
     private long lastRefreshMs;
     private long lastSuccessfulRefreshMs;
     private AuthenticationException authenticationException;
+    private KafkaException metadataException;
     private MetadataCache cache = MetadataCache.empty();
     private boolean needUpdate;
-    /* Topics with expiry time */
-    private final Map<String, Long> topics;
-    private final List<Listener> listeners;
     private final ClusterResourceListeners clusterResourceListeners;
-    private boolean needMetadataForAllTopics;
-    private final boolean allowAutoTopicCreation;
-    private final boolean topicExpiryEnabled;
     private boolean isClosed;
     private final Map<TopicPartition, Integer> lastSeenLeaderEpochs;
 
-    public Metadata(long refreshBackoffMs,
-                    long metadataExpireMs,
-                    boolean allowAutoTopicCreation) {
-        this(refreshBackoffMs, metadataExpireMs, allowAutoTopicCreation, false, new ClusterResourceListeners());
-    }
-
     /**
      * Create a new Metadata instance
-     * @param refreshBackoffMs The minimum amount of time that must expire between metadata refreshes to avoid busy
-     *        polling
-     * @param metadataExpireMs The maximum amount of time that metadata can be retained without refresh
-     * @param allowAutoTopicCreation If this and the broker config 'auto.create.topics.enable' are true, topics that
-     *                               don't exist will be created by the broker when a metadata request is sent
-     * @param topicExpiryEnabled If true, enable expiry of unused topics
+     *
+     * @param refreshBackoffMs         The minimum amount of time that must expire between metadata refreshes to avoid busy
+     *                                 polling
+     * @param metadataExpireMs         The maximum amount of time that metadata can be retained without refresh
+     * @param logContext               Log context corresponding to the containing client
      * @param clusterResourceListeners List of ClusterResourceListeners which will receive metadata updates.
      */
     public Metadata(long refreshBackoffMs,
                     long metadataExpireMs,
-                    boolean allowAutoTopicCreation,
-                    boolean topicExpiryEnabled,
+                    LogContext logContext,
                     ClusterResourceListeners clusterResourceListeners) {
+        this.log = logContext.logger(Metadata.class);
         this.refreshBackoffMs = refreshBackoffMs;
         this.metadataExpireMs = metadataExpireMs;
-        this.allowAutoTopicCreation = allowAutoTopicCreation;
-        this.topicExpiryEnabled = topicExpiryEnabled;
         this.lastRefreshMs = 0L;
         this.lastSuccessfulRefreshMs = 0L;
-        this.version = 0;
+        this.requestVersion = 0;
+        this.updateVersion = 0;
```

Comment:
I wonder if having the broker manage a metadata version would make this any easier?

Reply:
I think it would. It would be nice to have a simple version in the metadata response that we could use to stale metadata. I suspect the partition leader epoch is the closest we can get to that without introducing server metadata bottlenecks. That said, I think some amount of bookkeeping is unavoidable in the client since the set of topics that it is interested in changes over time. You need some way of telling whether pending metadata fetches reflect changes to the interested topic set.